src/platform/qt/library/LibraryController.cpp

```diff
@@ -0,0 +1,162 @@
+
+	m_libraryTree = new LibraryTree(this);
+	addWidget(m_libraryTree);
+	connect(m_libraryTree, &LibraryTree::StartGame, this, &LibraryController::startGame);
+
+	m_libraryGrid = new LibraryGrid(this);
+	addWidget(m_libraryGrid);
+	connect(m_libraryGrid, &LibraryGrid::StartGame, this, &LibraryController::startGame);
+
+	refresh();
+	selectLastBootedGame(); // TODO: doesn't belong here
+
+	setViewStyle(LibraryStyle::STYLE_TREE);
+}
+
+LibraryController::~LibraryController() {
+	mLibraryListingDeinit(&m_listing);
+	if (m_library) {
+		mLibraryDestroy(m_library);
+	}
+}
+
+void LibraryController::setViewStyle(LibraryStyle newStyle) {
+	m_currentStyle = newStyle;
+
+	AbstractGameList* newCurrentWidget = nullptr;
+	if (newStyle == LibraryStyle::STYLE_LIST || newStyle == LibraryStyle::STYLE_TREE) {
+		newCurrentWidget = m_libraryTree;
+	} else {
+		newCurrentWidget = m_libraryGrid;
+	}
+	newCurrentWidget->selectEntry(selectedEntry());
+	newCurrentWidget->setViewStyle(newStyle);
+	setCurrentWidget(dynamic_cast<QWidget*>(newCurrentWidget));
+}
+
+void LibraryController::selectEntry(LibraryEntry entry) {
+	dynamic_cast<AbstractGameList*>(currentWidget())->selectEntry(entry);
+}
+
+LibraryEntry LibraryController::selectedEntry() {
+	return dynamic_cast<AbstractGameList*>(currentWidget())->selectedEntry();
+}
+
+VFile* LibraryController::selectedVFile() {
+	LibraryEntry entry = selectedEntry();
+	if (entry) {
+		return mLibraryOpenVFile(m_library, entry->entry);
+	} else {
+		return nullptr;
+	}
+}
+
+QPair<QString, QString> LibraryController::selectedPath() {
+	LibraryEntry e = selectedEntry();
+	return e ? qMakePair(e->base(), e->filename()) : qMakePair<QString, QString>("", "");
+}
+
+void LibraryController::addDirectory(const QString& dir) {
+	mLibraryLoadDirectory(m_library, dir.toUtf8().constData());
+	refresh();
+}
+
+void LibraryController::setDirectory(const QString& dir) {
+	addDirectory(dir); // FIXME
+}
+
+void LibraryController::refresh() {
+	setDisabled(true);
+
+	QStringList allEntries;
+	QList<LibraryEntry> newEntries;
+
+	mLibraryListingClear(&m_listing);
+	mLibraryGetEntries(m_library, &m_listing, 0, 0, nullptr);
+	for (size_t i = 0; i < mLibraryListingSize(&m_listing); i++) {
+		mLibraryEntry* entry = mLibraryListingGetPointer(&m_listing, i);
+		QString fullpath = QString("%1/%2").arg(entry->base, entry->filename);
+		if (m_entries.contains(fullpath)) {
+			m_entries.value(fullpath)->entry = entry;
+		} else {
+			LibraryEntry libentry = QSharedPointer<CLibraryEntry>(new CLibraryEntry(entry));
+			m_entries.insert(fullpath, libentry);
+			newEntries.append(libentry);
+		}
+		allEntries.append(fullpath);
+	}
+
+	// Check for entries that were removed
+	QList<LibraryEntry> removedEntries;
+	for (QString& path : m_entries.keys()) {
+		if (!allEntries.contains(path)) {
+			removedEntries.append(m_entries.value(path));
+			m_entries.remove(path);
+		}
+	}
+
+	m_libraryTree->addEntries(newEntries);
+	m_libraryGrid->addEntries(newEntries);
```

Comment:
What would you suggest? Giving the GameList widgets access to m_entries wouldn't solve anything, because then they'd have to do the `for ...` to find the items added. "Seems" clunky, but there's not really any way this could fall over that I can see...

Additionally, this mechanism makes it possible to have the treed gamelist, which was quite well liked when Dolphin had it.